subprojects/core-api/src/main/java/org/gradle/api/artifacts/transform/ArtifactTransformOutputs.java

```diff
@@ -48,12 +48,12 @@
      *
      * @param output the output
      */
-    void registerOutputFile(File output);
+    void registerOutput(File output);
```

Comment:
The method Javadoc should say "file _or directory",_ and perhaps also mention that it's enough to register roots. I guess this is true for the `String`-taking version of this method, too.

Reply:
> This method needs be used when part of the {@link PrimaryInput} is the output of the transform.

I think this is assuming too much here, or is too prescriptive. We could say that you are allowed to use this method for example when filtering the primary input.

Reply:
We should also mention the order retaining quality of this call.

Reply:
I think registering roots is beside the point: We are not really registering an output but a result of the transform. The output is the workspace directory + the location of the result (written to a file).

Maybe we should rename all the methods to `registerResult*`?